fix+refactor(ua_client_connect, ua_discovery, server_register): elimi…

JianbinLiu-CFLab opened the pull request:
Hi jpfr，

**Problem**
During shutdown, LDS register/deregister emitted misleading errors (`BadInvalidState`, `BadInternalError`, etc.), polluting logs and sometimes triggering unwanted reconnect attempts.

**Changes**
* **Client** (`src/client/ua_client_connect.c`):
  * Treat shutdown teardown codes (`BadInvalidState`, `BadConnectionClosed`, `BadSecureChannelClosed`) as benign.
  * Preserve `connectStatus` on shutdown paths.
  * Trigger reconnect only when `oldState == CONNECTING && connectStatus == GOOD`.
  * Skip activity in `CLOSING` state.

* **Server Discovery** (`src/server/ua_discovery.c`):
  * Downgrade benign shutdown errors and skip reconnects during shutdown.
  * Downgrade `RegisterServer2` failure logs if shutting down.
  * In `UA_Server_deregisterDiscovery`, normalize teardown errors to `SUCCESS` and log at `INFO` level to prevent misleading errors.

* **Example** (`examples/discovery/server_register.c`):
  * Updated comments and tracing to avoid spurious shutdown errors.

**Validation**
* 100× register/deregister cycles:
  * **Before**: 43× `BadInvalidState` errors.
  * **After**: 0 errors.
* Shutdown now logs only INFO messages (e.g., “Discovery connection closing (BadInternalError) — treat as benign”).
* No change to runtime error handling.

jpfr commented on examples/discovery/server_register.c:
```diff
@@ -80,55 +94,73 @@ int main(int argc, char **argv) {
 
-    // Unregister the server from the discovery server.
-    memset(&cc, 0, sizeof(UA_ClientConfig));
-    UA_ClientConfig_setDefault(&cc);
```
You need to recreate the client config.
It gets cleared internally in the _register method.

This also impacts the cleanup below.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1688,37 +1689,48 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-        /* Set to closing (could be done already in UA_SecureChannel_shutdown).
-         * This impacts the handling of cancelled requests below. */
+        /* Mark SecureChannel as CLOSING (may already be done in UA_SecureChannel_shutdown).
+        * This affects how cancelled requests are handled below. */
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1688,37 +1689,48 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-         * this after setting the Session state. Otherwise we send out new Publish
-         * Requests immediately. */
+        /* Remove all outstanding asynchronous service calls — RequestIds are no longer valid.
+        * Do this after setting the Session state to avoid sending new PublishRequests. */
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1688,37 +1689,48 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-        /* The connection closed before it actually opened. Since we are
-         * connecting asynchronously, this happens when the TCP connection
-         * fails. Try to fall back on the initial EndpointUrl. */
+        /* If the connection closed before it actually opened (asynchronous connect failure),
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1688,37 +1689,48 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+        * attempt fallback to the initial EndpointUrl */
         if(oldState == UA_SECURECHANNELSTATE_CONNECTING &&
-           client->connectStatus == UA_STATUSCODE_GOOD)
+        client->connectStatus == UA_STATUSCODE_GOOD)
```
Revert to the old alignment.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1688,37 +1689,48 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-        /* Try to reconnect */
-        goto continue_connect;
+        /* Shutdown-aware behavior:
+        * - If we are in a normal reconnect scenario (oldState was CONNECTING and connectStatus is GOOD),
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-        if(client->channel.state != UA_SECURECHANNELSTATE_OPEN)
-            client->connectStatus = res;
+        /* If we are shutting down or the channel is already closing/closed,
+        * certain return codes are expected and should not be treated as errors. */
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                        UA_StatusCode_name(res));
+
+            /* Do NOT update connectStatus here. During shutdown this would only
+            * generate additional noisy state transitions without any benefit. */
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                        UA_StatusCode_name(res));
+
+            /* If processing the buffer fails before the SecureChannel has opened,
+            * then the client cannot recover. Reflect this in connectStatus so the
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+
+        if(benign_shutdown) {
+            UA_LOG_DEBUG(client->config.logging, UA_LOGCATEGORY_CLIENT,
+                        "Processing message during shutdown (%s) — treat as benign.",
```
Align the arguments within the parentheses.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+            * generate additional noisy state transitions without any benefit. */
+        } else {
+            UA_LOG_ERROR(client->config.logging, UA_LOGCATEGORY_CLIENT,
+                        "Processing the message returned the error code %s",
```
Align the arguments within the parentheses.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
-         *
-         * If the connectStatus is still good (the SecureChannel was fully
-         * opened before), then a reconnect is attempted. */
+        * Return immediately. notifyClientState will be called in the next
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/client/ua_client_connect.c:
```diff
@@ -1771,29 +1783,56 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
 
  continue_connect:
+     /* Safety check: do not proceed with connection activity if the channel
+     * is already closing. This avoids reconnect attempts during shutdown
```
Align the stars. Lines not longer than 80 characters.

jpfr commented on src/server/ua_discovery.c:
```diff
@@ -321,13 +322,33 @@ discoveryClientStateCallback(UA_Client *client,
-                         "Could not connect to the Discovery server with error %s",
-                         UA_StatusCode_name(connectStatus));
+        /* Treat typical shutdown/teardown codes as benign noise */
+        UA_Boolean benign_shutdown =
```
BADINTERNALERROR is not benign!
Please only filter on ar->shutdown.

jpfr reviewed (commented):
Very nice, this brings important improvements.
Check the inline comments.

Some comments are about cosmetics.
A few of the funky code changes look like a misconfigured editor...

Keep up the good job!

JianbinLiu-CFLab commented:
Hi jpfr,

I’ve updated the PR and split it into two commits as requested:

* `refactor(server_register.c): add Simple helpers and harden example`
* `fix(client_connect.c): avoid reconnect + error spam during graceful shutdown`

While reviewing the changes again, I realized I had made a mistake:
in `server_multicast.c` I added a `UA_ClientConfig_clear(&cc);` call which is not needed.
`UA_Server_deregisterDiscovery` already performs the cleanup internally.
I will correct this.

Question:
Would you prefer me to just fix this one line, or should I also unify the discovery examples so they consistently use the new *Simple helpers*?

Thanks a lot for pointing this out — I initially suspected the cleanup might be internal, but seeing the current API design was confusing.
Right now the API exposes only part of the `cc` lifecycle to the user, while `register` silently cleans up internally. This “half ownership” can be confusing, because it’s not clear who is responsible for what. Without looking into the internals, users could easily assume they need to call `UA_ClientConfig_clear` themselves and end up with a double-free.

Your feedback helped me realize this — I believe many users could run into the same trap.
Please advise how you would like me to proceed.
